Add Async Tuninterface

JettChenT opened the pull request.

conradev commented:
This looks great! Perhaps we could rename `write`/`read` to `send`/`recv`?

conradev commented:
We also need to disable the type on Windows:
```
error[E0432]: unresolved import `tokio::io::unix`
 --> tun-async\src\lib.rs:2:16
  |
2 | use tokio::io::unix::AsyncFd;
  |                ^^^^ could not find `unix` in `io`
```

I'm also open to moving `async` to being a submodule (and feature!) of the `tun` crate

JettChenT commented:
## Changes
- Added Tun AsyncInterface
- Removed the original `async` folder and made `async` a feature in the `tun` crate
- Renamed `read/write` to `send/recv` for both `TunInterface` and `AsyncTunInterface`.

conradev commented on .vscode/settings.json:
```diff
@@ -20,4 +20,7 @@
         "editor.defaultFormatter": "rust-lang.rust-analyzer",
     },
+    "rust-analyzer.linkedProjects": [
+        "./tun/Cargo.toml"
```
Is this needed? `rust-analyzer` should ideally pick up the `tun` crate from the Rust workspace

conradev replied:
It also looks like it disables auto-discovery:

> Disable project auto-discovery in favor of explicitly specified set of projects.

JettChenT replied:
Indeed! This is not needed, and it turns out to be a local development environment issue resulting in VSCode prompting me to add this.

conradev commented on tun/Cargo.toml:
```diff
@@ -11,12 +11,20 @@ socket2 = "0.4"
+futures = { version = "0.3.28", optional = true }
+
+[features]
+async = ["tokio/full", "dep:futures"]
```
We should only depend on the subset of tokio that we need, which might just be the `net` feature

conradev commented on tun/Cargo.toml:
```diff
@@ -11,12 +11,20 @@ socket2 = "0.4"
+futures = { version = "0.3.28", optional = true }
+
+[features]
+async = ["tokio/full", "dep:futures"]
```
I hate to ask for more changes, but because other asynchronous runtimes exist like `async-std`, perhaps we can name this feature (and its module) `tokio`?

JettChenT replied:
Done. Thought there would be a naming conflict issue if I named it `tokio`, but it turned out to be fine!

conradev reviewed (approved):
This is good to merge, minus the `.vscode/settings.json` change!

conradev commented on tun/src/tun_async/mod.rs:
```diff
@@ -0,0 +1,59 @@
+use std::io;
+use tokio::io::unix::AsyncFd;
+
+pub struct AsyncTunInterface {
```
I am okay with the `AsyncTunInterface` name, but there is precedent for naming the sync and async variants the same if we want to do that, like `tun::TunInterface` and `tun::tokio::TunInterface`

one example is reqwest: [`reqwest::blocking::Client`](https://docs.rs/reqwest/latest/reqwest/blocking/struct.Client.html) vs [`reqwest::Client`](https://docs.rs/reqwest/latest/reqwest/struct.Client.html)

JettChenT replied:
Hi, I do agree that it would be cleaner on the user side to follow this pattern, and renamed `AsyncTunInterface` to `TunInterface`.
